File: C++/202. Happy Number.cpp
#include <unordered_set>

using namespace std;

int square_sum(int n)
{
    int square = 0;

    while (n)
    {
        square += (n % 10) * (n % 10);
        n /= 10;
    }

    return square;
}

bool isHappy(int n)
{
    unordered_set<int> set;

    while (true)
    {
        if (n == 1)
            return true;
        set.insert(n);
        n = square_sum(n);

        if (set.find(n) != set.end())
            break;
    }
    return false;
}

bool isHappy(int slow)
{
    int fast = slow;

    do
    {
        slow = square_sum(slow);
        fast = square_sum(square_sum(fast));
    } while (slow != fast);

    return slow == 1;
}

File: Java/202. Happy Number.java

import java.util.HashSet;
import java.util.Set;

class Solution {
    /*
     * Time: O(n)
     * Space: O(n)
     */
    public boolean isHappy_usingSet(int n) {
        Set<Integer> set = new HashSet<>();

        while (n != 1) {
            n = findSquare(n);

            if (set.contains(n))
                return false;

            set.add(n);
        }

        return true;
    }

    private int findSquare(int n) {
        int result = 0;

        while (n > 0) {
            result += (n % 10 * n % 10);
            n /= 10;
        }

        return result;
    }

    public boolean isHappy(int n) {
        int fast = n, slow = n;

        do {
            slow = findSquare(slow);
            fast = findSquare(findSquare(fast));
        } while (slow != fast);

        return slow == 1;
    }
}